### PR TITLE
Fix #17863: `\yii\helpers\BaseInflector::slug()` doesn't work with an…

### LINKED ISSUE #17863: `\yii\helpers\BaseInflector::slug()` doesn't work with an empty string as a replacement argument.
### What steps will reproduce the problem?

```
Inflector::slug('text', '')
```
### What is the expected result?

```
text
```
### What do you get instead?

```
[yii\base\ErrorException] explode(): Empty delimiter
```

### Additional info

After [this commit](https://github.com/yiisoft/yii2/commit/1342eb718b6674c6ade3f2887dcf0710d2987e3b#diff-1d93dd129d9cbdd656d16f9f4716e24e) `\yii\helpers\BaseInflector::slug()` method cann't work with empty string as a `$replacement` argument. But it's needed for slugging with removing spaces.

May be it can be fixed by adding the checking like this:

```
if ($replacement) {
    $parts = explode($replacement, static::transliterate($string));
} else {
    $parts = [static::transliterate($string)];
}
```

| Q                | A
| ---------------- | ---
| Yii version      | >2.0.16?
| PHP version      | Any
| Operating system | Any

### PR DESCRIPTION
**Pull request by HaruAtari**
… empty string as a replacement argument.

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #17863

### CODE REVIEW

**Comment by HaruAtari** [comment]
Something strange. Tests have failed but all error are in the database tests. I didn't change any of them. How can I fix this, can anybody explain?

**Comment by kamarton** [comment]
```
There was 1 failure:

1) yiiunit\framework\ChangeLogTest::testContributorLine with data set #22 ('- Bug #17863: `\yii\helpers\B...atari)')
Failed asserting that '- Bug #17863: `\yii\helpers\BaseInflector::slug()` doesn't work with an empty string as a replacement argument. (haruatari)' matches PCRE pattern "/- (Bug|Enh|Chg|New)( #\d+(, #\d+)*)?(\s\(CVE-[\d-]+\))?: .*[^.] \(.+\)$/".
```

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -31,7 +31,7 @@ Yii Framework 2 Change Log
 - Bug #17803: Fix `ErrorHandler` unregister and register to only change global state when applicable (SamMousa)
 - Enh #17729: Path alias support was added to `yii\web\UploadedFile::saveAs()` (sup-ham)
 - Enh #17792: Add support for `aria` attributes to `yii\helpers\BaseHtml::renderTagAttributes()` (brandonkelly)
-
+- Bug #17863: `\yii\helpers\BaseInflector::slug()` doesn't work with an empty string as a replacement argument. (haruatari)
```
Comment (kamarton) [inline framework/CHANGELOG.md]: remove dot at end

File: framework/helpers/BaseInflector.php [hunk 2]
```diff
@@ -477,15 +477,21 @@ public static function tableize($className)
      */
     public static function slug($string, $replacement = '-', $lowercase = true)
     {
-        $parts = explode($replacement, static::transliterate($string));
+        if ($replacement) {
```
Comment (kamarton) [inline framework/helpers/BaseInflector.php]: `if ($prelacement === '')`

File: framework/helpers/BaseInflector.php [hunk 3]
```diff
@@ -477,15 +477,21 @@ public static function tableize($className)
...
+        }
 
         $replaced = array_map(function ($element) use ($replacement) {
             $element = preg_replace('/[^a-zA-Z0-9=\s—–-]+/u', '', $element);
             return preg_replace('/[=\s—–-]+/u', $replacement, $element);
         }, $parts);
 
         $string = trim(implode($replacement, $replaced), $replacement);
-        $string = preg_replace('#' . preg_quote($replacement) . '+#', $replacement, $string);
+        if ($replacement) {
```
Comment (kamarton) [inline framework/helpers/BaseInflector.php]: `if ($prelacement === '')`

**Comment by HaruAtari** [comment]
@samdark What should I do to fix database tests? How I understand it's something wrong with the test environment?

**Comment by simialbi** [comment]
@HaruAtari merge master branches changes and rerun the tests. The test environment got fixed

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thanks!

**Comment by LiuAndre** [comment]
Sorry for bothering, but the changelog is on the wrong version. 😐
It's appear on the 2.0.32, when it should be on 2.0.33
Sorry if my english is bad.

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2/commit/1d83073b121794736aa1f8771812a85ae05aa2d1

**Pull request merged**